File: 239-sliding-window-maximum/sliding-window-maximum.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        List<Integer> res = new ArrayList<>();
        Deque<Integer> q = new ArrayDeque<>();

        for (int i = 0; i < nums.length; i++) {
            while (!q.isEmpty() && q.getLast() < nums[i]) {
                q.removeLast();
            }
            q.addLast(nums[i]);

            if (i >= k && nums[i - k] == q.getFirst()) {
                q.removeFirst();
            }

            if (i >= k - 1) {
                res.add(q.getFirst());
            }
        }

        int[] resultArray = new int[res.size()];
        for (int i = 0; i < res.size(); i++) {
            resultArray[i] = res.get(i);
        }

        return resultArray;
    }
}

File: 239-sliding-window-maximum/sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        # result =[]

        # for i in range(len(nums)-k+1):
        #     cur_win = max(nums[i:i+k])
        #     result.append(cur_win)

        # return result

        q = deque()
        res = []

        for i, num in enumerate(nums):
            while q and q[-1] < num:
                q.pop()
            q.append(num)

            if i >=k and nums[i-k]==q[0]:
                q.popleft()

            if i >=k-1:
                res.append(q[0])

        return res